nifi_cluster_coordinator/services/environment_service.py

```diff
@@ -0,0 +1,176 @@
+                    return
+                update_version = True
+            except requests.exceptions.RequestException as exception:
+                logger.warning(f'Unable to temporarily delete version control from project: {project.name}, environment: {environment.name}.')
+                logger.warning(exception)
+                return
+
+        try:
+            put_url = '/' + url_helper.construct_path_parts(['process-groups', environment_json['id']])
+            response = requests.put(**cluster._get_connection_details(put_url), json=environment_json)
+            if response.status_code != 200:
+                logger.warning(response.text)
+                return
+            logger.info(f'Updated project: {project.name}, environment: {environment.name}.')
+            environment_json = response.json()
+        except requests.exceptions.RequestException as exception:
+            logger.warning(f'Unable to update project: {project.name}, environment: {environment.name}.')
+            logger.warning(exception)
+
+    if update_version:
+        _update_version(desired_version, cluster, project, environment, environment_json)
+
+
+def _delete(cluster: Cluster, project: Project, delete_environment_json):
+    logger = logging.getLogger(__name__)
+    environment_name = delete_environment_json['component']['name']
+
+    delete_url = '/' + url_helper.construct_path_parts(['process-groups', delete_environment_json['component']['id']])
+    try:
+        response = requests.delete(
+            **cluster._get_connection_details(delete_url),
+            params={'clientId': delete_environment_json['revision']['clientId'], 'version': str(delete_environment_json['revision']['version'])})
+        if response.status_code != 200:
+            logger.warning(response.text)
+            return
+        logger.info(f'Deleted project: {project.name}, environment: {environment_name}')
+    except requests.exceptions.RequestException as exception:
+        logger.warning(f'Unable to delete project: {project.name}, environment: {environment_name}.')
+        logger.warning(exception)
+
+
+def _update_version(desired_version, cluster: Cluster, project: Project, environment: ProjectEnvironment, environment_json):
+    logger = logging.getLogger(__name__)
+
+    registry_json = cluster.registeries_json_dict[project.registry_name]
+    version_update_json = {
+        'processGroupRevision': environment_json['revision'],
+        'versionControlInformation': {
+            'groupId': environment_json['id'],
+            'registryId': registry_json['id'],
+            'bucketId': project.bucket_id,
+            'flowId': project.flow_id,
+            'version': desired_version
+        }
+    }
+
+    # Todo: since version control is an async process inside nifi, we need an async logic or while loop to check
+    # the request status to issue delete request when the versionrequest is processed by the nifi server
```

Comment:
I think NiFi also garbage collects the update requests after a while.  So this may not be as big of an issue.  I'm not 100% sure on this though.